registry/repository.go

```diff
@@ -21,176 +20,123 @@ const (
-				return fmt.Errorf("failed to fetch manifest: %v: %v", desc.Digest, err)
-			}
-			if len(artifact.Blobs) == 0 {
-				continue
-			}
-			signatureManifests = append(signatureManifests, SignatureManifest{
-				Blob:        notationDescriptorFromArtifact(artifact.Blobs[0]),
-				Annotations: artifact.Annotations,
-			})
-		}
-		return nil
-	}); err != nil {
-		return nil, err
-	}
-	return signatureManifests, nil
+// ListSignatures returns signature manifests filtered by fn given the
+// artifact manifest descriptor
+func (c *repositoryClient) ListSignatures(ctx context.Context, desc ocispec.Descriptor, fn func(signatureManifests []ocispec.Descriptor) error) error {
+	return c.Repository.Referrers(ctx, desc, ArtifactTypeNotation, func(referrers []ocispec.Descriptor) error {
+		return fn(referrers)
+	})
 }
 
-// GetBlob downloads the content of the specified digest's Blob
-func (c *RepositoryClient) GetBlob(ctx context.Context, digest digest.Digest) ([]byte, error) {
-	desc, err := c.Repository.Blobs().Resolve(ctx, digest.String())
+// FetchSignatureBlob returns signature envelope blob and descriptor given
+// signature manifest descriptor
+func (c *repositoryClient) FetchSignatureBlob(ctx context.Context, desc ocispec.Descriptor) ([]byte, ocispec.Descriptor, error) {
+	sigManifest, err := c.getSignatureManifest(ctx, desc)
 	if err != nil {
-		return nil, err
+		return nil, ocispec.Descriptor{}, err
 	}
-	if desc.Size > maxBlobSizeLimit {
-		return nil, fmt.Errorf("signature blob too large: %d", desc.Size)
+	if len(sigManifest.Blobs) == 0 {
+		return nil, ocispec.Descriptor{}, errors.New("signature manifest missing signature envelope blob")
 	}
-	return content.FetchAll(ctx, c.Repository.Blobs(), desc)
+	sigDesc := ociDescriptorFromArtifact(sigManifest.Blobs[0])
+	if sigDesc.Size > maxBlobSizeLimit {
+		return nil, ocispec.Descriptor{}, fmt.Errorf("signature blob too large: %d", sigDesc.Size)
+	}
+	sigBlob, err := content.FetchAll(ctx, c.Repository.Blobs(), sigDesc)
+	if err != nil {
+		return nil, ocispec.Descriptor{}, err
+	}
+	return sigBlob, sigDesc, nil
 }
 
-// PutSignatureManifest creates and uploads an signature artifact linking the manifest and the signature
-func (c *RepositoryClient) PutSignatureManifest(ctx context.Context, signature []byte, signatureMediaType string, subjectManifest notation.Descriptor, annotations map[string]string) (notation.Descriptor, SignatureManifest, error) {
-	signatureDesc, err := c.uploadSignature(ctx, signature, signatureMediaType)
+// PushSignature creates and uploads an signature manifest along with its
+// linked signature envelope blob. Upon successful, PushSignature returns
+// signature envelope blob and manifest descriptors.
+func (c *repositoryClient) PushSignature(ctx context.Context, blob []byte, mediaType string, subject ocispec.Descriptor, annotations map[string]string) (blobDesc, manifestDesc ocispec.Descriptor, err error) {
+	blobDesc, err = c.uploadSignature(ctx, blob, mediaType)
```

Comment:
`updateSignature()` can be replaced by [oras.PushBytes()](https://pkg.go.dev/oras.land/oras-go/v2@main#PushBytes).